# 0. Import and Functions

## 0.1 Import 

In [122]:
import inflection


import pandas  as pd 
import seaborn as sns 
import plotly  as plt 
import numpy   as np 

In [123]:
df_raw = pd.read_csv("datas/WA_Fn-UseC_-HR-Employee-Attrition.csv")

## 0.2 Functions

# 1. Data Describe

In [124]:
df1 = df_raw.copy()

In [125]:
df1.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

## 1.1 Rename Columns

Explicação das colunas: 
<br>

* Age - Idade                         
* Attrition - Saída                 
* BusinessTravel - Viagem a Trabalho
* DailyRate - Diária                    
* Department - Departamento                  
* DistanceFromHome - Distância de Casa        
* Education - Educação
* EducationField - Campo de Educação     
* EmployeeCount  -  Contagem de Empregados
* EmployeeNumber -  Número de Empregado             
* EnvironmentSatisfaction - Satisfação da Empresa
* Gender - Gênero
* HourlyRate - Horas de Trabalho             
* JobInvolvement - Participação do Trbalho  
* JobLevel - Nível do Empregado                   
* JobRole  - Cargo                   
* JobSatisfaction - Satisfação do Trabalho            
* MaritalStatus   - Estado Civil
* MonthlyIncome   - Renda Mensal            
* MonthlyRate     - Taxa Mensal            
* NumCompaniesWorked  -  Número de Ex Empresas      
* Over18   -  Mais de 18 anos                   
* OverTime  - Ao longo do tempo                  
* PercentSalaryHike   - Percentual de Salário      
* PerformanceRating  - Avaliação de Desempenho     
* RelationshipSatisfaction   - Satisfação de Relacionamento    
* StandardHours - Horas Padrão              
* StockOptionLevel - Level das Ações           
* TotalWorkingYears - Anos trabalhados            
* TrainingTimesLastYear - Horas de Treinamento Ano Passado  
* WorkLifeBalance       - Equilíbrio Entre vida Pessoal   
* YearsAtCompany        - Anos de Empresa      
* YearsInCurrentRole    - Anos no mesmo cargo 
* YearsSinceLastPromotion - Anos desde a última promoção    
* YearsWithCurrManager    - Anos com gerente atual

In [126]:
cols_old = ['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager']

snakecase = lambda x: inflection.underscore( x )

cols_new = list( map( snakecase, cols_old ) )

# rename
df1.columns = cols_new

## 1.2 Data Dimensions

In [127]:
print( 'Number of Rows: {}'.format( df1.shape[0] ) )
print( 'Number of Cols: {}'.format( df1.shape[1] ) )

Number of Rows: 1470
Number of Cols: 35


## 1.3 Data Types

In [128]:
df1.dtypes

age                            int64
attrition                     object
business_travel               object
daily_rate                     int64
department                    object
distance_from_home             int64
education                      int64
education_field               object
employee_count                 int64
employee_number                int64
environment_satisfaction       int64
gender                        object
hourly_rate                    int64
job_involvement                int64
job_level                      int64
job_role                      object
job_satisfaction               int64
marital_status                object
monthly_income                 int64
monthly_rate                   int64
num_companies_worked           int64
over18                        object
over_time                     object
percent_salary_hike            int64
performance_rating             int64
relationship_satisfaction      int64
standard_hours                 int64
s

In [129]:
df1

age attrition    business_travel  daily_rate              department  \
0      41       Yes      Travel_Rarely        1102                   Sales   
1      49        No  Travel_Frequently         279  Research & Development   
2      37       Yes      Travel_Rarely        1373  Research & Development   
3      33        No  Travel_Frequently        1392  Research & Development   
4      27        No      Travel_Rarely         591  Research & Development   
...   ...       ...                ...         ...                     ...   
1465   36        No  Travel_Frequently         884  Research & Development   
1466   39        No      Travel_Rarely         613  Research & Development   
1467   27        No      Travel_Rarely         155  Research & Development   
1468   49        No  Travel_Frequently        1023                   Sales   
1469   34        No      Travel_Rarely         628  Research & Development   

      distance_from_home  education education_field  employee_count  \
0                      1          2   Life Sciences               1   
1                      8          1   Life Sciences               1   
2                      2          2           Other               1   
3                      3          4   Life Sciences               1   
4                      2          1         Medical               1   
...                  ...        ...             ...             ...   
1465                  23          2         Medical               1   
1466                   6          1         Medical               1   
1467                   4          3   Life Sciences               1   
1468                   2          3         Medical               1   
1469                   8          3         Medical               1   

      employee_number  ...  relationship_satisfaction standard_hours  \
0                   1  ...                          1             80   
1                   2  ...                          4             80   
2                   4  ...                          2             80   
3                   5  ...                          3             80   
4                   7  ...                          4             80   
...               ...  ...                        ...            ...   
1465             2061  ...                          3             80   
1466             2062  ...                          1             80   
1467             2064  ...                          2             80   
1468             2065  ...                          4             80   
1469             2068  ...                          1             80   

      stock_option_level  total_working_years  training_times_last_year  \
0                      0                    8                         0   
1                      1                   10                         3   
2                      0                    7                         3   
3                      0                    8                         3   
4                      1                    6                         3   
...                  ...                  ...                       ...   
1465                   1                   17                         3   
1466                   1                    9                         5   
1467                   1                    6                         0   
1468                   0                   17                         3   
1469                   0                    6                         3   

     work_life_balance  years_at_company years_in_current_role  \
0                    1                 6                     4   
1                    3                10                     7   
2                    3                 0                     0   
3                    3                 8                     7   
4                    3                 2                     2   
...                ...               ...              

## 1.5 Change DataTypes

In [130]:
df1["job_satisfaction"].value_counts()

4    459
3    442
1    289
2    280
Name: job_satisfaction, dtype: int64

In [131]:
df1["job_satisfaction1"] = df1["job_satisfaction"].apply(lambda x: "ruim"  if x == 1 else 
                                                       "regular" if x == 2 else 
                                                        "bom" if x == 3 else 
                                                        "excelente")

df1["performance_rating1"] = df1["performance_rating"].apply(lambda x: "ruim"  if x == 1 else 
                                                       "regular" if x == 2 else 
                                                        "bom" if x == 3 else 
                                                        "excelente")




## 1.6 Descriptive Statistics

In [132]:
num_attributes = df1.select_dtypes(exclude='object')
cat_attributes = df1.select_dtypes(include='object')

### 1.6.1 Numerical Atributes

In [133]:
num_attributes = num_attributes.drop(columns=['employee_number'], axis=1) 

describe = num_attributes.describe().T

describe['range'] = (num_attributes.max() - num_attributes.min()).tolist()
describe['variation coefficient'] = (num_attributes.std() / num_attributes.mean()).tolist()
describe['skew'] = num_attributes.skew().tolist()
describe['kurtosis'] = num_attributes.kurtosis().tolist()

describe

count          mean          std     min     25%  \
age                         1470.0     36.923810     9.135373    18.0    30.0   
daily_rate                  1470.0    802.485714   403.509100   102.0   465.0   
distance_from_home          1470.0      9.192517     8.106864     1.0     2.0   
education                   1470.0      2.912925     1.024165     1.0     2.0   
employee_count              1470.0      1.000000     0.000000     1.0     1.0   
environment_satisfaction    1470.0      2.721769     1.093082     1.0     2.0   
hourly_rate                 1470.0     65.891156    20.329428    30.0    48.0   
job_involvement             1470.0      2.729932     0.711561     1.0     2.0   
job_level                   1470.0      2.063946     1.106940     1.0     1.0   
job_satisfaction            1470.0      2.728571     1.102846     1.0     2.0   
monthly_income              1470.0   6502.931293  4707.956783  1009.0  2911.0   
monthly_rate                1470.0  14313.103401  7117.786044  2094.0  8047.0   
num_companies_worked        1470.0      2.693197     2.498009     0.0     1.0   
percent_salary_hike         1470.0     15.209524     3.659938    11.0    12.0   
performance_rating          1470.0      3.153741     0.360824     3.0     3.0   
relationship_satisfaction   1470.0      2.712245     1.081209     1.0     2.0   
standard_hours              1470.0     80.000000     0.000000    80.0    80.0   
stock_option_level          1470.0      0.793878     0.852077     0.0     0.0   
total_working_years         1470.0     11.279592     7.780782     0.0     6.0   
training_times_last_year    1470.0      2.799320     1.289271     0.0     2.0   
work_life_balance           1470.0      2.761224     0.706476     1.0     2.0   
years_at_company            1470.0      7.008163     6.126525     0.0     3.0   
years_in_current_role       1470.0      4.229252     3.623137     0.0     2.0   
years_since_last_promotion  1470.0      2.187755     3.222430     0.0     0.0   
years_with_curr_manager     1470.0      4.123129     3.568136     0.0     2.0   

                                50%       75%      max  range  \
age                            36.0     43.00     60.0     42   
daily_rate                    802.0   1157.00   1499.0   1397   
distance_from_home              7.0     14.00     29.0     28   
education                       3.0      4.00      5.0      4   
employee_count                  1.0      1.00      1.0      0   
environment_satisfaction        3.0      4.00      4.0      3   
hourly_rate                    66.0     83.75    100.0     70   
job_involvement                 3.0      3.00      4.0      3   
job_level                       2.0      3.00      5.0      4   
job_satisfaction                3.0      4.00      4.0      3   
monthly_income               4919.0   8379.00  19999.0  18990   
monthly_rate                14235.5  20461.50  26999.0  24905   
num_companies_worked            2.0      4.00      9.0      9   
percent_salary_hike            14.0     18.00     25.0     14   
performance_rating              3.0      3.00      4.0      1   
relationship_satisfaction       3.0      4.00      4.0      3   
standard_hours                 80.0     80.00     80.0      0   
stock_option_level              1.0      1.00      3.0      3   
total_working_years            10.0     15.00     40.0     40   
training_times_last_year        3.0      3.00      6.0      6   
work_life_balance               3.0      3.00      4.0      3   
years_at_company                5.0      9.00     40.0     40   
years_in_current_role           3.0      7.00     18.0     18   
years_since_last_promotion      1.0      3.00     15.0     15   
years_with_curr_manager         3.0      7.00     17.0     17   

                            variation coefficient      skew  kurtosis  
age                                      0.247411  0.413286 -0.404145  
daily_rate                               0.502824 -0.003519 -1.203823  
distance_from_h

* ANÁLISE 

**nenhum dos nossos dados são homogêneos. nenhum tem variation coefficient menor do que 15%**

age (idade): <br>
+ em média: 37 anos. Com skew positivo - concentração de dados para valores mais baixos. 

distancefromhome: <br> 
+ skew positiva, concentração para pessoas morando mais perto. Em média 9 km, um pouco longe. 


total_working_years: <br>
+ em média temos funcionários com 11 anos trabalhados número interessante


num_companies_worked: <br>
+ em média os colaboradores trabalharam em 3 empresas, porém o nosso coeficiente é muito alto isso demonstra uma alta variação de dados. 

stock_option_level: <br>
+ alta variação de dados 

### 1.6.2 Categorical Atributes

In [136]:
cat_attributes

attrition    business_travel              department education_field  \
0          Yes      Travel_Rarely                   Sales   Life Sciences   
1           No  Travel_Frequently  Research & Development   Life Sciences   
2          Yes      Travel_Rarely  Research & Development           Other   
3           No  Travel_Frequently  Research & Development   Life Sciences   
4           No      Travel_Rarely  Research & Development         Medical   
...        ...                ...                     ...             ...   
1465        No  Travel_Frequently  Research & Development         Medical   
1466        No      Travel_Rarely  Research & Development         Medical   
1467        No      Travel_Rarely  Research & Development   Life Sciences   
1468        No  Travel_Frequently                   Sales         Medical   
1469        No      Travel_Rarely  Research & Development         Medical   

      gender                   job_role marital_status over18 over_time  \
0     Female            Sales Executive         Single      Y       Yes   
1       Male         Research Scientist        Married      Y        No   
2       Male      Laboratory Technician         Single      Y       Yes   
3     Female         Research Scientist        Married      Y       Yes   
4       Male      Laboratory Technician        Married      Y        No   
...      ...                        ...            ...    ...       ...   
1465    Male      Laboratory Technician        Married      Y        No   
1466    Male  Healthcare Representative        Married      Y        No   
1467    Male     Manufacturing Director        Married      Y       Yes   
1468    Male            Sales Executive        Married      Y        No   
1469    Male      Laboratory Technician        Married      Y        No   

     job_satisfaction1 performance_rating1  
0            excelente                 bom  
1              regular           excelente  
2                  bom                 bom  
3                  bom                 bom  
4              regular                 bom  
...                ...                 ...  
1465         excelente                 bom  
1466              ruim                 bom  
1467           regular           excelente  
1468           regular                 bom  
1469               bom                 bom  

[1470 rows x 11 columns]

In [146]:
cat_attributes['job_satisfaction1'].value_counts(normalize=True) 

excelente    0.312245
bom          0.300680
ruim         0.196599
regular      0.190476
Name: job_satisfaction1, dtype: float64

In [135]:
cat_attributes['performance_rating1'].value_counts() 

bom          1244
excelente     226
Name: performance_rating1, dtype: int64

In [149]:
cat_attributes['attrition'].value_counts(normalize = True) 

No     0.838776
Yes    0.161224
Name: attrition, dtype: float64

In [148]:
cat_attributes['business_travel'].value_counts(normalize=True) 

Travel_Rarely        0.709524
Travel_Frequently    0.188435
Non-Travel           0.102041
Name: business_travel, dtype: float64

In [147]:
cat_attributes['department'].value_counts(normalize=True) 

Research & Development    0.653741
Sales                     0.303401
Human Resources           0.042857
Name: department, dtype: float64

In [141]:
cat_attributes['education_field'].value_counts() 

Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: education_field, dtype: int64

In [142]:
cat_attributes['gender'].value_counts() 

Male      882
Female    588
Name: gender, dtype: int64

In [143]:
cat_attributes['marital_status'].value_counts() 

Married     673
Single      470
Divorced    327
Name: marital_status, dtype: int64

In [144]:
cat_attributes['job_role'].value_counts() 

Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: job_role, dtype: int64

** INTEPRETAÇÃO ** 
* job_satisfaction1 - temos maior concentração em bom e excelente 
* performance_rating1 -  não temos dados diferentes de bom e excelente 
* department - Research & Development é o maior concentração de 65% da amostra 
* business_travel - quase 70% viajam raramente 
* attrition - só temos 16% dos co